stdlib/json.gr

```diff
@@ -0,0 +1,2308 @@
+ * Following examples have whitespaces and line breaks replaced with visible
+ * charactes for illustrative purposes.
+ * 
+ * `NoIndentation`
+ * ```
+ * {↵
+ * "currency":·"€",↵
+ * "price":·99.9↵
+ * }
+ * ```
+ * 
+ * `IndentWithTab`
+ * ```
+ * {↵
+ * →"currency":·"€",↵
+ * →"price":·99.9↵
+ * }
+ * ```
+ * 
+ * `IndentWithSpaces(2)`
+ * ```
+ * {↵
+ * ··"currency":·"€",↵
+ * ··"price":·99.9↵
+ * }
+ * ```
+ * 
+ * `IndentWithSpaces(4)`
+ * ```
+ * {↵
+ * ····"currency":·"€",↵
+ * ····"price":·99.9↵
+ * }
+ * ```
+ */
+export enum IndentationFormat {
+  NoIndentation,
+  IndentWithTab,
+  IndentWithSpaces(Number),
+}
+
+/**
+ * Controls how arrays are printed in custom formatting.
+ * 
+ * Following examples have whitespaces and line breaks replaced with visible
+ * charactes for illustrative purposes.
+ * 
+ * `CompactArrayEntries`
+ * ```
+ * []
+ * ```
+ * 
+ * ```
+ * [1]
+ * ```
+ * 
+ * ```
+ * [1,2,3]
+ * ```
+ * 
+ * `OneArrayEntryPerLine`
+ * ```
+ * []
+ * ```
+ * 
+ * ```
+ * [↵
+ * ··1↵
+ * ]
+ * ```
+ * 
+ * ```
+ * [↵
+ * ··1,↵
+ * ··2,↵
+ * ··3↵
+ * ]
+ * ```
+ */
+export enum ArrayFormat {
+  CompactArrayEntries,
+  OneArrayEntryPerLine,
+}
+
+/**
+ * Controls how objects are printed in custom formatting.
+ * 
+ * Following examples have whitespaces and line breaks replaced with visible
+ * charactes for illustrative purposes.
+ * 
+ * `CompactObjectEntries`
+ * ```
+ * {}
+ * ```
+ * 
+ * ```
+ * {"a":1}
+ * ```
+ * 
+ * ```
+ * {"a":1,"b":2,"c":3}
+ * ```
+ * 
+ * `OneObjectEntryPerLine`
+ * ```
+ * {}
+ * ```
+ * 
+ * ```
+ * {↵
+ * ··"a":·1↵
+ * }
+ * ```
+ * 
+ * ```
+ * {↵
+ * ··"a":·1,↵
+ * ··"b":·2,↵
+ * ··"c":·3↵
+ * }
+ * ```
+ */
+export enum ObjectFormat {
+  CompactObjectEntries,
+  OneObjectEntryPerLine,
+}
+
+/**
+ * Controls line ending type in custom formatting.
+ */
+export enum LineEnding {
+  NoLineEnding,
+  LineFeed,
+  CarriageReturnLineFeed,
+  CarriageReturn,
+}
+
+/**
+ * Allows fine grained control of formatting in JSON printing.
+ */
+export record FormattingSettings {
+  indentation: IndentationFormat,
+  arrayFormat: ArrayFormat,
+  objectFormat: ObjectFormat,
+  lineEnding: LineEnding,
+  finishWithNewLine: Bool,
+  escapeAllControlPoints: Bool,
+  escapeHTMLUnsafeSequences: Bool,
+  escapeNonASCII: Bool,
+}
+
+/**
+ * Compact formatting that minimizes the size of resulting JSON at cost of not
+ * being easily human readable.
+ * 
+ * Only performs minimal string escaping as required by the ECMA-404 standard,
+ * so the result needs to be treated as proper unicode and is not safe to be
+ * transported in ASCII encoding.
+ * 
+ * The following example have whitespaces, line breaks and control points
+ * replaced with visible characters.
+ * ```
+ * {"currency":"€","price":99.9,"currencyDescription":"EURO␡","script·unembeddable":"You·cannot·pase·a·JSON·object·directly·into·a·<script>·tag·in·HTML·if·it·contains·</·with·unescaped·forward·slash"}
+ * ```
+ */
+export let defaultCompactFormat = () =>
```

Comment:
I'm almost wondering if we should expose this API as some sort of
```
enum Format {
  Compact,
  Pretty,
  PrettyAndSafe,
  CompactAndSafe,
  Custom(FormattingSettings)
}
```
and update the user-facing functions to use those instead. @cician @ospencer thoughts?

Reply:
I'm fine with this change, but should I keep both `toString` and `toStringCompact`? The API would feel much cleaner with just `toString`, especially if we had optional parameters. One thing that worries me a bit is that would probably make it so the non compact printing function probably could not be discarded by tree shaking anymore. For many use cases pretty printing is not needed or even undesired.

Reply:
If we keep separate compact/pretty APIs, I believe they should be reversed. e.g. `toString` should be compact and `toPrettyString` (or something) should be the prettyprinter. Mostly because sending things across any boundary will want to be compact.